File: Data-Structures-and-Algorithms-using-PHP/String/TST.php
<?php
class Node {
	public $data;
	public $isLastChar;
	public $left;
	public $equal;
	public $right;
	Function __construct($d){
		$this->data = $d;
		$this->left = NULL;
		$this->equal = NULL;
		$this->right = NULL;
		$this->isLastChar = false;
	}
}
class TST {
	protected $root;
	Function __construct() {
		$this->root = NULL;
	}
	public Function insert_String($word) {
		$this->root = $this->insert_Node_String_I ( $this->root, $word, 0 );
	}
	protected Function insert_Node_String_I($curr, $word, $wordIndex) {
		if ($curr == NULL)
			$curr = new Node( $word [$wordIndex] );
		if ($word [$wordIndex] < $curr->data)
			$curr->left = $this->insert_Node_String_I ( $curr->left, $word, $wordIndex );
		else if ($word [$wordIndex] > $curr->data)
			$curr->right = $this->insert_Node_String_I ( $curr->right, $word, $wordIndex );
		else {
			if ($wordIndex < (strlen ( $word ) - 1))
				$curr->equal = $this->insert_Node_String_I ( $curr->equal, $word, ($wordIndex + 1) );
			else
				$curr->isLastChar = true;
		}
		return $curr;
	}
	protected Function Find_Node_String_I($curr, $word, $wordIndex) {
		if ($curr == NULL)
			return false;
		if ($word [$wordIndex] < $curr->data)
			return $this->Find_Node_String_I ( $curr->left, $word, $wordIndex );
		else if ($word [$wordIndex] > $curr->data)
			return $this->Find_Node_String_I ( $curr->right, $word, $wordIndex );
		else {
			if ($wordIndex == (strlen ( $word ) - 1))
				return $curr->isLastChar;
			
			return $this->Find_Node_String_I ( $curr->equal, $word, ($wordIndex + 1) );
		}
	}
	public Function Find_String($word) {
		$ret = $this->Find_Node_String_I ( $this->root, $word, 0 );
		echo ("$word :: ");
		if ($ret)
			echo ("Found <br/>");
		else
			echo ("Not Found<br/>");
		return $ret;
	}
}

$tt = new TST();
$tt->insert_String ( "banana" );
$tt->insert_String ( "apple" );
$tt->insert_String ( "mango" );
echo ("\nSearch results For apple, banana, grapes and mango :<br/>");
$tt->Find_String ( "apple" );
$tt->Find_String ( "banana" );
$tt->Find_String ( "mango" );
$tt->Find_String ( "grapes" );
?>

File: Data-Structures-and-Algorithms-using-Swift/String/TST.swift
class TST {
	class Node  {
		var data : Character = "a"
		var isLastChar : Bool = false
		var left : Node? = nil
		var equal : Node? = nil
		var right : Node? = nil
	}	
	
	var root : Node? = nil

	func insert(_ word : String) {
		let str : [Character] = Array(word.lowercased())
		self.root = self.insertUtil(node : self.root, word : str, wordIndex : 0)
	}
	
	func insertUtil(node : Node?, word : [Character], wordIndex : Int) -> Node {
		var curr : Node
		if node == nil {
			curr = Node()
			curr.data = word[wordIndex]
		} else {
			curr = node!
		}
		if word[wordIndex] < curr.data {
			curr.left = self.insertUtil(node : curr.left, word : word, wordIndex : wordIndex)
		} else if word[wordIndex] > curr.data {
			curr.right = self.insertUtil(node : curr.right, word : word, wordIndex : wordIndex)
		} else {
			if wordIndex < word.count-1 {
				curr.equal = self.insertUtil(node : curr.equal, word : word, wordIndex : wordIndex+1)
			} else {
				curr.isLastChar = true
			}
		}
		return curr
	}
	
	func find(_ word : String) -> Bool {
		let str : [Character] = Array(word.lowercased())
		return self.findUtil(curr : self.root, word : str, wordIndex : 0)
	}

	func findUtil(curr : Node?, word : [Character], wordIndex : Int) -> Bool {
		guard let curr = curr else {
			return false
		}
	
		if word[wordIndex] < curr.data {
			return self.findUtil(curr : curr.left, word : word, wordIndex : wordIndex)
		} else if word[wordIndex] > curr.data {
			return self.findUtil(curr : curr.right, word : word, wordIndex : wordIndex)
		} else {
			if wordIndex == word.count-1 {
				return curr.isLastChar
			}
			return self.findUtil(curr : curr.equal, word : word, wordIndex : wordIndex+1)
		}
	}
}

// Testing code.
let tt = TST()
tt.insert("banana")
tt.insert("apple")
tt.insert("mango")
print("Apple Found :", tt.find("apple"))
print("Banana Found :", tt.find("banana"))
print("Grapes Found :", tt.find("grapes"))
